File: events/2017/02/02.py
from itertools  import permutations
from re         import split

def preprocessing(input_):
    lines = [[int(n) for n in split(r'\s', line)] for line in input_.splitlines()]
    return lines

def solver(spreadsheet):
    checksum = 0
    checkdiv = 0
    
    for row in spreadsheet:
        checksum += max(row) - min(row)
        for a, b in permutations(row, 2):
            if a % b == 0: 
                checkdiv += a // b
                break

    yield checksum
    yield checkdiv

File: Events/2017/02/02.rs
use itertools::MinMaxResult;

aoc::main!();

fn preprocessing(input: &str) -> Vec<Vec<u16>> {
    input
    .lines()
    .map(|row| 
        row
        .split_whitespace()
        .map(|n| n.parse().unwrap())
        .collect_vec())
    .collect()
}

fn part_1(spreadsheet: Vec<Vec<u16>> ) -> u16 {
    spreadsheet
    .iter()
    .map(|row| row.iter().minmax())
    .map(|result| 
        match result {
            MinMaxResult::MinMax(a, b) => b - a ,
            _ => unreachable!()})
    .sum()
}

fn part_2(spreadsheet: Vec<Vec<u16>>) -> u16 {
    spreadsheet
    .iter()
    .map(|row| 
        row
        .iter()
        .permutations(2)
        .find(|pair| pair[0] % pair[1] == 0)
        .unwrap())
    .map(|pair,| pair[0] / pair[1])
    .sum()
}